marimo/_save/hash.py

```diff
@@ -4,106 +4,246 @@
+from marimo._utils.variables import (
+    get_cell_from_local,
+    if_local_then_mangle,
+    unmangle_local,
+)
 
 if TYPE_CHECKING:
+    from hashlib import _Hash as Hash
     from types import CodeType
 
     from marimo._ast.cell import CellId_t, CellImpl
+    from marimo._runtime.context.types import RuntimeContext
     from marimo._runtime.dataflow import DirectedGraph
     from marimo._save.loaders import Loader
 
-BASE_PRIMITIVES = (str, numbers.Number, type(None))
+    # Union[list, torch.Tensor, jax.numpy.ndarray,
+    #             np.ndarray, scipy.sparse.spmatrix]
+    Tensor = Any
 
 
-def hash_module(code: Optional[CodeType]) -> bytes:
+DEFAULT_HASH = "sha256"
+
+
+def hash_module(
+    code: Optional[CodeType], hash_type: str = DEFAULT_HASH
+) -> bytes:
+    hash_alg = hashlib.new(hash_type, usedforsecurity=False)
     if not code:
-        # SHA256 hash of 32 zero bytes, in the case of no code object
+        # Hash of zeros, in the case of no code object as a recognizable noop.
         # Artifact of typing for mypy, but reasonable fallback.
-        return b"0" * 32
-
-    sha = hashlib.sha256()
+        return b"0" * len(hash_alg.digest())
 
     def process(code_obj: CodeType) -> None:
         # Recursively hash the constants that are also code objects
         for const in code_obj.co_consts:
             if isinstance(const, types.CodeType):
                 process(const)
             else:
-                sha.update(str(const).encode("utf8"))
+                hash_alg.update(str(const).encode("utf8"))
         # Concatenate the names and bytecode of the current code object
         # Will cause invalidation of variable naming at the top level
-        sha.update(bytes("|".join(code_obj.co_names), "utf8"))
-        sha.update(code_obj.co_code)
+        hash_alg.update(bytes("|".join(code_obj.co_names), "utf8"))
+        hash_alg.update(code_obj.co_code)
 
     process(code)
-    return sha.digest()
+    return hash_alg.digest()
 
 
-def hash_raw_module(module: ast.Module) -> bytes:
+def hash_raw_module(
+    module: ast.Module, hash_type: str = DEFAULT_HASH
+) -> bytes:
     # AST has to be compiled to code object prior to process.
     return hash_module(
         compile(
             module,
             "<hash>",
             mode="exec",
             flags=ast.PyCF_ALLOW_TOP_LEVEL_AWAIT,
-        )
+        ),
+        hash_type,
     )
 
 
-def hash_cell_impl(cell: CellImpl) -> bytes:
-    return hash_module(cell.body) + hash_module(cell.last_expr)
+def hash_cell_impl(cell: CellImpl, hash_type: str = DEFAULT_HASH) -> bytes:
+    return hash_module(cell.body, hash_type) + hash_module(
+        cell.last_expr, hash_type
+    )
 
 
-def build_execution_hash(
-    graph: DirectedGraph, cell_id: CellId_t
-) -> Optional[ValidCacheSha]:
+def hash_and_dequeue_execution_refs(
+    hash_alg: Hash,
+    cell_id: CellId_t,
+    graph: DirectedGraph,
+    refs: set[str],
+) -> None:
```

Comment:
Can you add a function docstring? It looks like this mutates `refs` and `hash_alg`?

Reply:
Good point. Carrying around a lot of state, probably better just to make a `BlockHash` class